libs/template/renderer.go

```diff
@@ -0,0 +1,260 @@
+package template
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"path"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/databricks/cli/libs/filer"
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/databricks-sdk-go/logger"
+	"golang.org/x/exp/slices"
+)
+
+type inMemoryFile struct {
+	// Unix like path for the file (using '/' as the separator)
+	path    string
+	content []byte
+}
+
+// Renders a databricks template as a project
+type renderer struct {
+	ctx context.Context
+
+	// A config that is the "dot" value available to any template being rendered.
+	// Refer to https://pkg.go.dev/text/template for how templates can use
+	// this "dot" value
+	config map[string]any
+
+	// A base template with helper functions and user defined template in the
+	// library directory loaded. This is used as the base to compute any project
+	// templates during file tree walk
+	baseTemplate *template.Template
+
+	// List of in memory files generated from template
+	files []*inMemoryFile
+
+	// Glob patterns for files and directories to skip. There are three possible
+	// outcomes for skip:
+	//
+	// 1. File is not generated. This happens if one of the file's parent directories
+	// match a glob pattern
+	//
+	// 2. File is generated but not persisted to disk. This happens if the file itself
+	// matches a glob pattern, but none of it's parents match a glob pattern from the list
+	//
+	// 3. File is persisted to disk. This happens if the file and it's parent directories
+	// do not match any glob patterns from this list
+	skipPatterns []string
+
+	// Filer rooted at template root. The file tree from this root is walked to
+	// generate the project
+	templateFiler filer.Filer
+
+	// Filer rooted at the project instance root. Allows writing files generated
+	// when walking the template file tree
+	instanceFiler filer.Filer
+}
+
+func newRenderer(ctx context.Context, config map[string]any, libraryRoot, instanceRoot, templateRoot string) (*renderer, error) {
+	// Initialize new template, with helper functions loaded
+	tmpl := template.New("").Funcs(HelperFuncs)
+
+	// Load user defined associated templates from the library root
+	libraryGlob := filepath.Join(libraryRoot, "*")
+	matches, err := filepath.Glob(libraryGlob)
+	if err != nil {
+		return nil, err
+	}
+	if len(matches) != 0 {
+		tmpl, err = tmpl.ParseGlob(libraryGlob)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	templateFiler, err := filer.NewLocalClient(templateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceFiler, err := filer.NewLocalClient(instanceRoot)
```

Comment:
The filers don't support chmod's. We don't need this abstraction here, strictly speaking, because we only deal with local file paths. Wouldn't it be better to just use `filepath.Join` and `os.WriteFile` when writing?

Reply:
done